File: src/simulation/ai/memory/WorldSummary.js
import { GROUP_IDS } from "../../../config/aiConfig.js";
import { TILE } from "../../../config/constants.js";
import {
  GROUP_DEFAULT_STATE,
  listGroupStates,
  listGroupTransitions,
} from "../../npc/state/StateGraph.js";
import { getScenarioRuntime } from "../../../world/scenarios/ScenarioFactory.js";
import { MIN_FOOD_FOR_GROWTH } from "../../population/PopulationGrowthSystem.js";
import {
  sampleTerrainAggregates,
  sampleSoilAggregates,
  sampleNodeDepletionCounts,
  sampleWaterConnectivity,
} from "../colony/ColonyPerceiver.js";

const POLICY_GROUP_ORDER = Object.freeze([
  GROUP_IDS.WORKERS,
  GROUP_IDS.TRADERS,
  GROUP_IDS.SABOTEURS,
  GROUP_IDS.HERBIVORES,
  GROUP_IDS.PREDATORS,
]);

function countPassability(grid) {
  let passable = 0;
  for (let i = 0; i < grid.tiles.length; i += 1) {
    if (grid.tiles[i] !== TILE.WALL && grid.tiles[i] !== TILE.WATER) passable += 1;
  }
  return passable / grid.tiles.length;
}

function estimateCongestion(state) {
  let moving = 0;
  let idle = 0;
  for (const e of [...state.agents, ...state.animals]) {
    const speed = Math.hypot(e.vx, e.vz);
    if (speed > 0.2) moving += 1;
    else idle += 1;
  }
  const total = moving + idle;
  if (total === 0) return 0;
  return idle / total;
}

function resolveDominantState(stateCounts, fallbackState) {
  let bestState = fallbackState;
  let bestCount = -1;
  for (const [state, count] of Object.entries(stateCounts ?? {})) {
    const safe = Number(count) || 0;
    if (safe > bestCount) {
      bestCount = safe;
      bestState = state;
    }
  }
  return bestState;
}

function buildTransitionHints(groupId, world, groupStats) {
  const dominant = groupStats.dominantState;
  const avgHunger = Number(groupStats.avgHunger ?? 0.5);
  const food = Number(world.resources?.food ?? 0);
  const threatSignals = Number(world.events?.length ?? 0);
  const hints = [];

  if (groupId === GROUP_IDS.WORKERS) {
    if (food < MIN_FOOD_FOR_GROWTH || avgHunger < 0.34) hints.push("seek_food -> eat -> seek_task");
    if ((groupStats.carrying ?? 0) > Math.max(4, groupStats.count * 0.5)) hints.push("deliver -> seek_task");
    if (dominant === "wander" || dominant === "idle") hints.push("wander -> seek_task -> harvest");
  } else if (groupId === GROUP_IDS.TRADERS) {
    if (avgHunger < 0.36) hints.push("seek_food -> eat -> seek_trade");
    if (dominant === "wander" || dominant === "idle") hints.push("wander -> seek_trade");
    if (threatSignals > 0) hints.push("trade -> seek_trade (short ttl)");
  } else if (groupId === GROUP_IDS.SABOTEURS) {
    if (avgHunger < 0.34) hints.push("seek_food -> eat -> scout");
    hints.push("scout -> sabotage -> evade -> scout");
  } else if (groupId === GROUP_IDS.HERBIVORES) {
    if (dominant === "flee") hints.push("flee -> regroup -> graze");
    else hints.push("graze -> regroup -> wander");
  } else if (groupId === GROUP_IDS.PREDATORS) {
    if (dominant === "rest") hints.push("rest -> stalk -> hunt");
    else hints.push("stalk -> hunt -> feed -> roam");
  }

  return Array.from(new Set(hints)).slice(0, 3);
}

function buildFrontierSummary(runtime) {
  return {
    connectedRoutes: Number(runtime.connectedRoutes ?? 0),
    totalRoutes: Number(runtime.routes?.length ?? 0),
    readyDepots: Number(runtime.readyDepots ?? 0),
    totalDepots: Number(runtime.depots?.length ?? 0),
    brokenRouteCount: Number((runtime.routes ?? []).filter((route) => !route.connected).length),
    brokenRoutes: (runtime.routes ?? []).filter((route) => !route.connected).map((route) => route.label).slice(0, 3),
    readyDepotLabels: (runtime.depots ?? []).filter((depot) => depot.ready).map((depot) => depot.label).slice(0, 3),
    unreadyDepotCount: Number((runtime.depots ?? []).filter((depot) => !depot.ready).length),
    unreadyDepots: (runtime.depots ?? []).filter((depot) => !depot.ready).map((depot) => depot.label).slice(0, 3),
  };
}

function buildOperationsSummary(world) {
  const frontier = world.frontier ?? {};
  const logistics = world.logistics ?? {};
  const recovery = world.gameplay?.recovery ?? {};
  const ecology = world.ecology ?? {};
  const issues = [];

  if ((frontier.brokenRoutes ?? []).length > 0) issues.push(`repair ${frontier.brokenRoutes[0]}`);
  if ((frontier.unreadyDepots ?? []).length > 0) issues.push(`reclaim ${frontier.unreadyDepots[0]}`);
  if (Number(logistics.overloadedWarehouses ?? 0) > 0) issues.push("relieve depot congestion");
  if (Number(logistics.strandedCarryWorkers ?? 0) > 0) issues.push("unstick delivery paths");
  if (Number(ecology.pressuredFarms ?? 0) > 0) issues.push("respond to farm pressure");
  if (Number(recovery.collapseRisk ?? 0) >= 60) issues.push("preserve recovery window");

  return {
    keyIssues: issues.slice(0, 5),
    focus: issues[0] ?? "maintain stable frontier throughput",
  };
}

export function buildWorldSummary(state) {
  const workers = state.agents.filter((a) => a.type === "WORKER").length;
  const visitors = state.agents.filter((a) => a.type === "VISITOR").length;
  const herbivores = state.animals.filter((a) => a.kind === "HERBIVORE").length;
  const predators = state.animals.filter((a) => a.kind === "PREDATOR").length;
  const runtime = getScenarioRuntime(state);
  const currentObjective = state.gameplay?.objectives?.[state.gameplay?.objectiveIndex ?? 0] ?? null;

  const summary = {
    simTimeSec: Math.floor(state.metrics.timeSec),
    resources: {
      food: Number(state.resources.food.toFixed(2)),
      wood: Number(state.resources.wood.toFixed(2)),
      stone: Number(Number(state.resources?.stone ?? 0).toFixed(2)),
      herbs: Number(Number(state.resources?.herbs ?? 0).toFixed(2)),
      meals: Number(Number(state.resources?.meals ?? 0).toFixed(2)),
      medicine: Number(Number(state.resources?.medicine ?? 0).toFixed(2)),
      tools: Number(Number(state.resources?.tools ?? 0).toFixed(2)),
    },
    population: { workers, visitors, herbivores, predators },
    buildings: { ...state.buildings },
    scenario: {
      id: String(runtime.scenario?.id ?? ""),
      title: String(runtime.scenario?.title ?? ""),
      family: String(runtime.scenario?.family ?? ""),
      summary: String(runtime.scenario?.summary ?? ""),
    },
    objective: currentObjective
      ? {
          id: currentObjective.id,
          title: currentObjective.title,
          description: currentObjective.description,
          progress: Number(Number(currentObjective.progress ?? 0).toFixed(1)),
          hint: String(state.gameplay?.objectiveHint ?? ""),
        }
      : {
          id: "",
          title: "All objectives completed",
          description: "",
          progress: 100,
          hint: String(state.gameplay?.objectiveHint ?? ""),
        },
    gameplay: {
      doctrine: String(state.gameplay?.doctrine ?? "balanced"),
      prosperity: Number(Number(state.gameplay?.prosperity ?? 0).toFixed(2)),
      threat: Number(Number(state.gameplay?.threat ?? 0).toFixed(2)),
      doctrineMastery: Number(Number(state.gameplay?.doctrineMastery ?? 1).toFixed(3)),
      recovery: {
        charges: Number(state.gameplay?.recovery?.charges ?? 0),
        activeBoostSec: Number(Number(state.gameplay?.recovery?.activeBoostSec ?? 0).toFixed(1)),
        collapseRisk: Number(Number(state.gameplay?.recovery?.collapseRisk ?? 0).toFixed(1)),
        lastReason: String(state.gameplay?.recovery?.lastReason ?? ""),
      },
    },
    frontier: buildFrontierSummary(runtime),
    weather: {
      current: state.weather.current,
      timeLeftSec: Number(state.weather.timeLeftSec.toFixed(1)),
      pressureScore: Number((state.weather.pressureScore ?? 0).toFixed(2)),
      hazardFronts: Number(state.weather.hazardFronts?.length ?? 0),
      hazardFocusSummary: String(state.weather.hazardFocusSummary ?? ""),
    },
    traffic: {
      congestion: Number(estimateCongestion(state).toFixed(3)),
      passableRatio: Number(countPassability(state.grid).toFixed(3)),
    },
    logistics: {
      carryingWorkers: Number(state.metrics?.logistics?.carryingWorkers ?? 0),
      totalCarryInTransit: Number(Number(state.metrics?.logistics?.totalCarryInTransit ?? 0).toFixed(2)),
      avgDepotDistance: Number(Number(state.metrics?.logistics?.avgDepotDistance ?? 0).toFixed(2)),
      strandedCarryWorkers: Number(state.metrics?.logistics?.strandedCarryWorkers ?? 0),
      overloadedWarehouses: Number(state.metrics?.logistics?.overloadedWarehouses ?? 0),
      busiestWarehouseLoad: Number(state.metrics?.logistics?.busiestWarehouseLoad ?? 0),
      stretchedWorksites: Number(state.metrics?.logistics?.stretchedWorksites ?? 0),
      isolatedWorksites: Number(state.metrics?.logistics?.isolatedWorksites ?? 0),
      summary: String(state.metrics?.logistics?.summary ?? "Logistics: unavailable"),
    },
    ecology: {
      pressuredFarms: Number(state.metrics?.ecology?.pressuredFarms ?? 0),
      maxFarmPressure: Number(Number(state.metrics?.ecology?.maxFarmPressure ?? 0).toFixed(2)),
      frontierPredators: Number(state.metrics?.ecology?.frontierPredators ?? 0),
      migrationHerds: Number(state.metrics?.ecology?.migrationHerds ?? 0),
      hotspotFarms: (state.metrics?.ecology?.hotspotFarms ?? []).slice(0, 3),
      summary: String(state.metrics?.ecology?.summary ?? "Ecology: unavailable"),
    },
    events: state.events.active.map((e) => ({
      type: e.type,
      status: e.status,
      intensity: e.intensity,
      targetLabel: String(e.payload?.targetLabel ?? ""),
      severity: String(e.payload?.severity ?? ""),
      pressure: Number(Number(e.payload?.pressure ?? 0).toFixed(2)),
      contestedTiles: Number(e.payload?.contestedTiles ?? 0),
      remainingSec: Number(Math.max(0, e.durationSec - e.elapsedSec).toFixed(1)),
    })),
    spatialPressure: {
      weatherPressure: Number((state.metrics.spatialPressure?.weatherPressure ?? 0).toFixed(2)),
      eventPressure: Number((state.metrics.spatialPressure?.eventPressure ?? 0).toFixed(2)),
      contestedZones: Number(state.metrics.spatialPressure?.contestedZones ?? 0),
      activeEventCount: Number(state.metrics.spatialPressure?.activeEventCount ?? 0),
    },
    aiMode: state.ai.mode,

    // v0.8.2: terrain, soil, node depletion counts, water connectivity
    terrain: sampleTerrainAggregates(state.grid),
    soil: sampleSoilAggregates(state.grid),
    nodes: sampleNodeDepletionCounts(state.grid),
    connectivity: sampleWaterConnectivity(state),
  };

  summary.operations = buildOperationsSummary(summary);

  if (state.ai?.strategy) {
    summary._strategyContext = {
      priority: state.ai.strategy.priority,
      resourceFocus: state.ai.strategy.resourceFocus,
      defensePosture: state.ai.strategy.defensePosture,
      riskTolerance: state.ai.strategy.riskTolerance,
      workerFocus: state.ai.strategy.workerFocus,
      environmentPreference: state.ai.strategy.environmentPreference,
    };
  }

  return summary;
}

export function buildPolicySummary(state) {
  const byGroup = {};

  for (const a of state.agents) {
    const stateNode = a.blackboard?.fsm?.state ?? a.stateLabel ?? "idle";
    byGroup[a.groupId] ??= { count: 0, avgHunger: 0, carrying: 0, states: {} };
    byGroup[a.groupId].count += 1;
    byGroup[a.groupId].avgHunger += a.hunger ?? 0;
    byGroup[a.groupId].carrying += (a.carry?.food ?? 0) + (a.carry?.wood ?? 0);
    byGroup[a.groupId].states[stateNode] = (byGroup[a.groupId].states[stateNode] ?? 0) + 1;
  }

  for (const a of state.animals) {
    const stateNode = a.blackboard?.fsm?.state ?? a.stateLabel ?? "idle";
    byGroup[a.groupId] ??= { count: 0, states: {} };
    byGroup[a.groupId].count += 1;
    byGroup[a.groupId].states[stateNode] = (byGroup[a.groupId].states[stateNode] ?? 0) + 1;
  }

  for (const g of Object.values(byGroup)) {
    if (g.count && g.avgHunger !== undefined) {
      g.avgHunger = Number((g.avgHunger / g.count).toFixed(3));
    }
  }

  const world = buildWorldSummary(state);
  const transitionContext = {};
  for (const groupId of POLICY_GROUP_ORDER) {
    const stats = byGroup[groupId] ?? { count: 0, avgHunger: 0, carrying: 0, states: {} };
    const dominantState = resolveDominantState(stats.states, GROUP_DEFAULT_STATE[groupId] ?? "idle");
    transitionContext[groupId] = {
      count: Number(stats.count ?? 0),
      avgHunger: Number(stats.avgHunger ?? 0),
      carrying: Number(stats.carrying ?? 0),
      states: stats.states ?? {},
      dominantState,
      stateNodes: listGroupStates(groupId),
      transitions: listGroupTransitions(groupId),
      preferredPaths: buildTransitionHints(groupId, world, {
        ...stats,
        dominantState,
      }),
    };
  }

  return {
    world,
    groups: byGroup,
    stateTransitions: {
      groups: transitionContext,
      generatedAtSec: Number(state.metrics.timeSec ?? 0),
    },
  };
}

File: project_utopia/simulation/ai/memory/world_summary.py
"""WorldSummary — observation-envelope builder for the LLM channels.

Python port of ``WorldSummary.js``. The full JS implementation reads from
``ScenarioFactory``, ``PopulationGrowthSystem``, ``ColonyPerceiver`` and the
versioned ``Grid`` — none of which are in scope for this Phase 1 subagent.
We therefore expose a **structural** ``build_world_summary`` that produces
the same camelCase output shape but reads optional/nested fields defensively
from the passed-in state dict. Subagents porting the substrate (Phase 2) can
swap in the rich helpers when the dependencies land.

Field names are kept camelCase on the wire to stay drop-in compatible with
the JS NDJSON traces.
"""

from __future__ import annotations

import math
from typing import Any, Iterable

POLICY_GROUP_ORDER = ("workers", "traders", "saboteurs", "herbivores", "predators")
GROUP_DEFAULT_STATE = {
    "workers": "idle",
    "traders": "idle",
    "saboteurs": "scout",
    "herbivores": "graze",
    "predators": "stalk",
}


def _num(value: Any, default: float = 0.0) -> float:
    try:
        v = float(value)
    except (TypeError, ValueError):
        return default
    if not math.isfinite(v):
        return default
    return v


def _round(value: Any, ndigits: int = 2, default: float = 0.0) -> float:
    return round(_num(value, default), ndigits)


def _list_of(value: Any) -> list[Any]:
    return list(value) if isinstance(value, list) else []


def _resolve_dominant_state(state_counts: dict[str, Any], fallback: str) -> str:
    best_state = fallback
    best_count = -1
    for state_name, count in (state_counts or {}).items():
        safe = int(_num(count))
        if safe > best_count:
            best_count = safe
            best_state = state_name
    return best_state


def build_world_summary(state: dict[str, Any]) -> dict[str, Any]:
    """Build the deterministic world summary dict (camelCase keys on the wire)."""
    if not isinstance(state, dict):
        state = {}
    metrics = state.get("metrics") or {}
    resources = state.get("resources") or {}
    buildings = state.get("buildings") or {}
    gameplay = state.get("gameplay") or {}
    recovery = gameplay.get("recovery") or {}
    weather = state.get("weather") or {}
    events_state = state.get("events") or {}
    ai_state = state.get("ai") or {}

    agents: Iterable[dict[str, Any]] = _list_of(state.get("agents"))
    animals: Iterable[dict[str, Any]] = _list_of(state.get("animals"))
    workers = sum(1 for a in agents if a.get("type") == "WORKER")
    visitors = sum(1 for a in agents if a.get("type") == "VISITOR")
    herbivores = sum(1 for a in animals if a.get("kind") == "HERBIVORE")
    predators = sum(1 for a in animals if a.get("kind") == "PREDATOR")

    objective = None
    objectives = gameplay.get("objectives") or []
    obj_idx = int(_num(gameplay.get("objectiveIndex")))
    if isinstance(objectives, list) and 0 <= obj_idx < len(objectives):
        objective = objectives[obj_idx]

    summary = {
        "simTimeSec": int(_num(metrics.get("timeSec"))),
        "resources": {
            "food": _round(resources.get("food")),
            "wood": _round(resources.get("wood")),
            "stone": _round(resources.get("stone")),
            "herbs": _round(resources.get("herbs")),
            "meals": _round(resources.get("meals")),
            "medicine": _round(resources.get("medicine")),
            "tools": _round(resources.get("tools")),
        },
        "population": {
            "workers": workers,
            "visitors": visitors,
            "herbivores": herbivores,
            "predators": predators,
            "foodHeadroomSec": _round(
                (state.get("population") or {}).get("foodHeadroomSec", 9999), 1, 9999
            ),
        },
        "buildings": dict(buildings),
        "objective": (
            {
                "id": str(objective.get("id", "")),
                "title": str(objective.get("title", "")),
                "description": str(objective.get("description", "")),
                "progress": _round(objective.get("progress"), 1),
                "hint": str(gameplay.get("objectiveHint", "")),
            }
            if isinstance(objective, dict)
            else {
                "id": "",
                "title": "All objectives completed",
                "description": "",
                "progress": 100,
                "hint": str(gameplay.get("objectiveHint", "")),
            }
        ),
        "gameplay": {
            "doctrine": str(gameplay.get("doctrine", "balanced")),
            "prosperity": _round(gameplay.get("prosperity")),
            "threat": _round(gameplay.get("threat")),
            "doctrineMastery": _round(gameplay.get("doctrineMastery", 1.0), 3, 1.0),
            "recovery": {
                "charges": int(_num(recovery.get("charges"))),
                "activeBoostSec": _round(recovery.get("activeBoostSec"), 1),
                "collapseRisk": _round(recovery.get("collapseRisk"), 1),
                "lastReason": str(recovery.get("lastReason", "")),
            },
        },
        "weather": {
            "current": str(weather.get("current", "clear")),
            "timeLeftSec": _round(weather.get("timeLeftSec"), 1),
            "pressureScore": _round(weather.get("pressureScore"), 2),
            "hazardFronts": len(_list_of(weather.get("hazardFronts"))),
            "hazardFocusSummary": str(weather.get("hazardFocusSummary", "")),
        },
        "frontier": dict(state.get("frontier") or {}),
        "logistics": dict((metrics.get("logistics") or {})),
        "ecology": dict((metrics.get("ecology") or {})),
        "events": [
            {
                "type": e.get("type"),
                "status": e.get("status"),
                "intensity": _round(e.get("intensity"), 2),
                "targetLabel": str(((e.get("payload") or {}).get("targetLabel")) or ""),
                "severity": str(((e.get("payload") or {}).get("severity")) or ""),
                "pressure": _round((e.get("payload") or {}).get("pressure"), 2),
            }
            for e in _list_of(events_state.get("active"))
        ],
        "spatialPressure": {
            "weatherPressure": _round(
                (metrics.get("spatialPressure") or {}).get("weatherPressure"), 2
            ),
            "eventPressure": _round(
                (metrics.get("spatialPressure") or {}).get("eventPressure"), 2
            ),
            "contestedZones": int(
                _num((metrics.get("spatialPressure") or {}).get("contestedZones"))
            ),
            "activeEventCount": int(
                _num((metrics.get("spatialPressure") or {}).get("activeEventCount"))
            ),
        },
        "aiMode": (ai_state.get("mode") or "off"),
    }

    summary["operations"] = _build_operations_summary(summary)

    strategy = ai_state.get("strategy")
    if isinstance(strategy, dict):
        summary["_strategyContext"] = {
            "priority": strategy.get("priority"),
            "resourceFocus": strategy.get("resourceFocus"),
            "defensePosture": strategy.get("defensePosture"),
            "riskTolerance": strategy.get("riskTolerance"),
            "workerFocus": strategy.get("workerFocus"),
            "environmentPreference": strategy.get("environmentPreference"),
        }
    return summary


def _build_operations_summary(world: dict[str, Any]) -> dict[str, Any]:
    frontier = world.get("frontier") or {}
    logistics = world.get("logistics") or {}
    recovery = (world.get("gameplay") or {}).get("recovery") or {}
    ecology = world.get("ecology") or {}
    issues: list[str] = []
    broken = _list_of(frontier.get("brokenRoutes"))
    if broken:
        issues.append(f"repair {broken[0]}")
    unready = _list_of(frontier.get("unreadyDepots"))
    if unready:
        issues.append(f"reclaim {unready[0]}")
    if _num(logistics.get("overloadedWarehouses")) > 0:
        issues.append("relieve depot congestion")
    if _num(logistics.get("strandedCarryWorkers")) > 0:
        issues.append("unstick delivery paths")
    if _num(ecology.get("pressuredFarms")) > 0:
        issues.append("respond to farm pressure")
    if _num(recovery.get("collapseRisk")) >= 60:
        issues.append("preserve recovery window")
    return {
        "keyIssues": issues[:5],
        "focus": issues[0] if issues else "maintain stable frontier throughput",
    }


def build_policy_summary(state: dict[str, Any]) -> dict[str, Any]:
    """Build the per-group policy summary (groups + state-transition context)."""
    by_group: dict[str, dict[str, Any]] = {}
    for a in _list_of(state.get("agents")):
        fsm = a.get("fsm") or {}
        state_node = (
            (fsm.get("state") if isinstance(fsm, dict) else None)
            or ((a.get("blackboard") or {}).get("fsm") or {}).get("state")
            or a.get("stateLabel")
            or "idle"
        )
        state_node = str(state_node).lower()
        group = a.get("groupId") or "workers"
        bucket = by_group.setdefault(group, {"count": 0, "avgHunger": 0.0, "carrying": 0, "states": {}})
        bucket["count"] += 1
        bucket["avgHunger"] += float(a.get("hunger") or 0.0)
        carry = a.get("carry") or {}
        bucket["carrying"] += float(carry.get("food") or 0.0) + float(carry.get("wood") or 0.0)
        bucket["states"][state_node] = bucket["states"].get(state_node, 0) + 1

    for a in _list_of(state.get("animals")):
        fsm = a.get("fsm") or {}
        state_node = (
            (fsm.get("state") if isinstance(fsm, dict) else None)
            or ((a.get("blackboard") or {}).get("fsm") or {}).get("state")
            or a.get("stateLabel")
            or "idle"
        )
        state_node = str(state_node).lower()
        group = a.get("groupId") or "herbivores"
        bucket = by_group.setdefault(group, {"count": 0, "states": {}})
        bucket["count"] += 1
        bucket["states"][state_node] = bucket["states"].get(state_node, 0) + 1

    for bucket in by_group.values():
        count = bucket.get("count") or 0
        if count and "avgHunger" in bucket:
            bucket["avgHunger"] = round(bucket["avgHunger"] / count, 3)

    world = build_world_summary(state)
    transition_context: dict[str, Any] = {}
    for group_id in POLICY_GROUP_ORDER:
        stats = by_group.get(group_id) or {"count": 0, "avgHunger": 0.0, "carrying": 0, "states": {}}
        dominant = _resolve_dominant_state(stats.get("states") or {}, GROUP_DEFAULT_STATE.get(group_id, "idle"))
        transition_context[group_id] = {
            "count": int(stats.get("count") or 0),
            "avgHunger": float(stats.get("avgHunger") or 0.0),
            "carrying": float(stats.get("carrying") or 0.0),
            "states": dict(stats.get("states") or {}),
            "dominantState": dominant,
        }
    return {
        "world": world,
        "groups": by_group,
        "stateTransitions": {
            "groups": transition_context,
            "generatedAtSec": float(_num((state.get("metrics") or {}).get("timeSec"))),
        },
    }


__all__ = [
    "GROUP_DEFAULT_STATE",
    "POLICY_GROUP_ORDER",
    "build_policy_summary",
    "build_world_summary",
]
